Firestore/core/src/remote/firebase_metadata_provider_apple.mm

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/remote/firebase_metadata_provider_apple.h"
+
+#import "FirebaseCore/Sources/Private/FIRAppInternal.h"
+#import "FirebaseCore/Sources/Private/FIRHeartbeatInfo.h"
+#import "FirebaseCore/Sources/Private/FIROptionsInternal.h"
+
+#include "Firestore/core/src/util/string_apple.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+namespace {
+
+using util::MakeString;
+
+std::string GetUserAgent() {
+  return MakeString([FIRApp firebaseUserAgent]);
+}
+
+std::string GetHeartbeat() {
+  return std::to_string([FIRHeartbeatInfo heartbeatCodeForTag:@"fire-fst"]);
```

Comment:
FYI, currently `[FIRHeartbeatInfo heartbeatCodeForTag:]` method performs file operations to read and update data in the heartbeat storage. Ideally this call should not happen on the main thread. Also it may be worth to check if the current implementation provides performance good enough for Firestore use case. 

Please let me know if you have any questions or concerns.

Reply:
Firestore runs all networking and and I/O operations on a dedicated queue. This can only ever be called from that queue so this won't have a chance to run on the main thread.